teamcity.build

```diff
@@ -245,231 +246,278 @@
+				att.OwnerElement.RemoveAttribute(att.Name);
+		}
+
+		public static void StripElements(XmlDocument source, string xpath)
+		{
+			foreach (XmlElement el in source.SelectNodes(xpath))
+				el.ParentNode.RemoveChild(el);
+		}
+
+		[Function("StripTimings")]
+		public static string StripTimings(string testResultFile, string outputFile)
+		{
+			try
+			{
+				XmlDocument testResults = new XmlDocument();
+				testResults.Load(testResultFile);
+
+				StripElements(testResults, "//stack-trace");
+				StripAttributes(testResults, "//@cwd");
+				StripAttributes(testResults, "//@total");
+				StripAttributes(testResults, "//@date");
+				StripAttributes(testResults, "//@time");
+				StripAttributes(testResults, "//@machine-name");
+				StripAttributes(testResults, "//@user");
+				StripAttributes(testResults, "//@user-domain");
+
+				outputFile = Path.GetDirectoryName(testResultFile) + "/" + outputFile;
+				testResults.Save(outputFile);
+				return outputFile;
+			}
+			catch (Exception e)
+			{
+				Console.WriteLine(e);
+				throw;
+			}
+		}
+
+		public enum ResultStatus
+		{
+			None,
+			Ignored,
+			Success,
+			Failure
+		}
+
+		public class Result
+		{
+			public string Name;
+			public ResultStatus Status;
+			public bool Executed;
+			public bool Success;
+
+			public bool Inconclusive
+			{
+				get
+				{
+					return Executed && !Success && Status == ResultStatus.Success;
+				}
+			}
+
+			private static Result Parse(XmlNode testCase)
+			{
+				var result = new Result
+				{
+					Name = testCase.Attributes["name"].Value,
+					Executed = bool.Parse(testCase.Attributes["executed"].Value)
+				};
+
+				ResultStatus status;
+				Enum.TryParse<ResultStatus>(testCase.Attributes["result"].Value, out status);
+				result.Status = status;
+
+				if (result.Executed)
+					result.Success = bool.Parse(testCase.Attributes["success"].Value);
+
+				return result;
+			}
+
+			public override string ToString()
+			{
+				var prefix = "IGNORED - ";
+				if (Executed)
+					prefix = Success ? "PASS    - " :
+						Inconclusive ? "INCONCL - " : "FAIL    - ";
+				return prefix + Name;
+			}
+
+			public static IList<Result> ParseFile(string file)
+			{
+				var run1 = new XmlDocument();
+				run1.Load(file);
+
+				var results = new List<Result>();
+				foreach (XmlNode testCase in run1.SelectNodes("//test-case"))
+				{
+					var result = Parse(testCase);
+					results.Add(result);
+				}
+				return results;
+			}
+		}
+
+		[Function("CompareResults")]
+		public static string CompareResults(string currentResult, string lastResult)
+		{
+			try
+			{
+				var outputFile = Path.GetDirectoryName(currentResult) + "/Comparison.txt";
+				var report = new StringBuilder();
+
+				report.AppendLine("Comparison Results");
+				report.AppendLine("==================");
+
+				var before = Result.ParseFile(lastResult);
+				var beforeByName = before.GroupBy(b => b.Name).ToDictionary(g => g.Key, g => g.First());
```

Comment:
Some existing result file have duplicated tests. The previous code was matching new tests to each of them, with those new tests ending up potentially duplicated in the results lists below.

Here I simply take the first. The dup test is currently always `NHibernate.Test.Linq.WhereTests.CanUseCompareInQuery(p => (p.ShippingWeight.CompareTo(4.98) <= 0),17,False)`. I am going to fix its test case, it does it twice. All outcome were same, but just in case, il will change the logic above for considering non-failing results first.